From: Borislav Petkov <bp@suse.de>
Date: Sun, 17 Dec 2017 16:01:57 +0100
Subject: x86/spec_ctrl: Check whether IBPB is enabled before using it
Patch-mainline: submitted on 2018/1/9
References: bsc#1068032

Check whether IBPB is enabled before using it.

Carved out from a patch by Tim Chen <tim.c.chen@linux.intel.com>

Signed-off-by: Borislav Petkov <bp@suse.de>
---
 arch/x86/include/asm/spec_ctrl.h | 3 ++-
 arch/x86/kernel/cpu/spec_ctrl.c  | 7 +++++++
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/arch/x86/include/asm/spec_ctrl.h b/arch/x86/include/asm/spec_ctrl.h
index 372f68d6a6f7..384b3b8b79b3 100644
--- a/arch/x86/include/asm/spec_ctrl.h
+++ b/arch/x86/include/asm/spec_ctrl.h
@@ -95,10 +95,11 @@ void x86_enable_ibrs(void);
 void x86_disable_ibrs(void);
 void stuff_RSB(void);
 unsigned int x86_ibrs_enabled(void);
+unsigned int x86_ibpb_enabled(void);
 
 static inline void x86_ibp_barrier(void)
 {
-	if (static_cpu_has(X86_FEATURE_SPEC_CTRL))
+	if (x86_ibpb_enabled())
 		native_wrmsrl(MSR_IA32_PRED_CMD, FEATURE_SET_IBPB);
 }
 
diff --git a/arch/x86/kernel/cpu/spec_ctrl.c b/arch/x86/kernel/cpu/spec_ctrl.c
index 1475d51f9c4f..672c32da1b02 100644
--- a/arch/x86/kernel/cpu/spec_ctrl.c
+++ b/arch/x86/kernel/cpu/spec_ctrl.c
@@ -12,6 +12,7 @@
  * Keep it open for more flags in case needed.
  */
 static unsigned int ibrs_state = 0;
+static unsigned int ibpb_state = 0;
 
 unsigned int notrace x86_ibrs_enabled(void)
 {
@@ -19,6 +20,12 @@ unsigned int notrace x86_ibrs_enabled(void)
 }
 EXPORT_SYMBOL_GPL(x86_ibrs_enabled);
 
+unsigned int notrace x86_ibpb_enabled(void)
+{
+	return ibpb_state;
+}
+EXPORT_SYMBOL_GPL(x86_ibpb_enabled);
+
 void x86_disable_ibrs(void)
 {
 	if (x86_ibrs_enabled())

